File: php/ShippingsettingsSample.php
<?php
/**
 * Copyright 2016 Google Inc. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

require_once 'BaseSample.php';

/**
 * Class for running through some example interactions with the
 * Shippingsettings service. Provides some public methods that
 * can be used to make your own calls if desired.
 */
class ShippingsettingsSample extends BaseSample {

  public function run() {
    printf("Retrieving shipping settings for %s\n", $this->merchantId);
    $oldSettings = $this->getShippingsettings($this->merchantId);
    $this->printShippingsettings($oldSettings);
    printf("Updating shipping settings for %s\n", $this->merchantId);
    $newSettings = $this->createShippingSample();
    $this->updateShippingsettings($this->merchantId, $newSettings);
    $this->printShippingsettings($this->getShippingsettings($this->merchantId));
    printf("Replacing old shipping settings for %s\n", $this->merchantId);
    $this->updateShippingsettings($this->merchantId, $oldSettings);
    $this->printShippingsettings($this->getShippingsettings($this->merchantId));
  }

  /**
   * Retrieves the shipping settings for {@code $accountId}. If
   * the configuration specifies a non-multi-client account, then the caller
   * can only retrieve the own information for that account, else they can
   * ask for the information of subaccounts.
   *
   * @param int $accountId the Merchant Center ID of the account for which
   *     to retrieve information
   * @return Shippingsettings the Shippingsettings resource that contains
   *     the shipping information for the requested account
   * @throws InvalidArgumentException if a non-MCA requests information for
   *     a different account
   */
  public function getShippingsettings($accountId) {
    if ($accountId != $this->merchantId) {
      $this->mustBeMCA('Non-multi-client accounts can only get their own '
          . 'information.');
    }
    $settings = $this->service->shippingsettings->get(
        $this->merchantId, $accountId);
    return $settings;
  }

  /**
   * Updates the shipping settings for {@code $accountId} using the
   * information in {@code $settings}.  If the configuration specifies
   * a non-multi-client account, then the caller can only update that
   * account's information, else they can update the information of subaccounts.
   *
   * @param int $accountId the Merchant Center ID of the account for which
   *     to retrieve information
   * @param Shippingsettings $settings the Shippingsettings resource with
   *     which to update the account
   * @throws InvalidArgumentException if a non-MCA requests information for
   *     a different account
   */
  public function updateShippingsettings($accountId, $settings) {
    if ($accountId != $this->merchantId) {
      $this->mustBeMCA('Non-multi-client accounts can only set their own '
          . 'information.');
      return;
    }
    $this->service->shippingsettings->update(
        $this->merchantId, $accountId, $settings);
  }

  /**
   * Prints the shipping information contained in {@code $settings}.
   *
   * @param Shippingsettings $settings the Shippingsettings resource to print
   */
  public function printShippingsettings($settings) {
    printf("Information for account %s:\n", $settings->getAccountId());
    if (empty($settings->getPostalCodeGroups())) {
      print "- No postal code groups defined.\n";
    } else {
      printf("- There are %d postal code group(s):\n",
          count($settings->getPostalCodeGroups()));
      foreach ($settings->getPostalCodeGroups() as $group) {
        printf("  Postal group "%s":\n", $group->getName());
        printf("  - Country: %s\n", $group->getCountry());
        printf("  - Contains %d postal code range(s).\n",
            count($group->getPostalCodeRanges()));
      }
    }
    if (empty($settings->getServices())) {
      print "- No shipping services defined.\n";
    } else {
      printf("- There are %d shipping service(s):\n",
          count($settings->getServices()));
      foreach ($settings->getServices() as $service) {
        printf("  Service "%s":\n", $service->getName());
        printf("  - Active: %s\n", $service->getActive() ? 'yes' : 'no');
        printf("  - Country: %s\n", $service->getDeliveryCountry());
        printf("  - Currency: %s\n", $service->getCurrency());
        printf("  - Delivery time: %d - %d days\n",
          $service->getDeliveryTime()->getMinTransitTimeInDays(),
          $service->getDeliveryTime()->getMaxTransitTimeInDays());
        printf("  - Contains %d rate group(s).\n",
          count($service->getRateGroups()));
      }
    }
  }

  private function createShippingSample() {
    $deliveryTime = new Google_Service_ShoppingContent_DeliveryTime();
    $deliveryTime->setMinTransitTimeInDays(3);
    $deliveryTime->setMaxTransitTimeInDays(7);

    $price = new Google_Service_ShoppingContent_Price();
    $price->setValue(5.00);
    $price->setCurrency('USD');

    $value = new Google_Service_ShoppingContent_Value();
    $value->setFlatRate($price);

    $rateGroup = new Google_Service_ShoppingContent_RateGroup();
    $rateGroup->setSingleValue($value);

    $service = new Google_Service_ShoppingContent_Service();
    $service->setName('USPS');
    $service->setActive(true);
    $service->setDeliveryCountry('US');
    $service->setCurrency('USD');
    $service->setDeliveryTime($deliveryTime);
    $service->setRateGroups([$rateGroup]);

    $settings = new Google_Service_ShoppingContent_ShippingSettings();
    $settings->setPostalCodeGroups([]);
    $settings->setServices([$service]);

    return $settings;
  }
}

$sample = new ShippingsettingsSample();
$sample->run();

File: dotnet/source/ShippingsettingsSample.cs
using System;
using Google.Apis.ShoppingContent.v2_1;
using Google.Apis.ShoppingContent.v2_1.Data;
using System.Collections.Generic;

namespace ShoppingSamples.Content
{
    /// <summary>
    /// A sample consumer that runs multiple requests against the Shippingsettings
    /// service in the Content API for Shopping.
    /// <para>These include:
    /// <list type="bullet">
    /// <item>
    /// <description>Shippingsettings.get</description>
    /// </item>
    /// <item>
    /// <description>Shippingsettings.update</description>
    /// </item></list></para>
    /// </summary>
    public class ShippingsettingsSample
    {
        private ShoppingContentService service;

        /// <summary>Initializes a new instance of the <see cref="ShippingsettingsSample"/> class.</summary>
        /// <param name="service">Content service object on which to run the requests.</param>
        public ShippingsettingsSample(ShoppingContentService service)
        {
            this.service = service;
        }

        /// <summary>Runs requests against the Content API for Shopping.</summary>
        internal void RunCalls(ulong merchantId)
        {
            ShippingSettings oldSettings = GetShippingSettings(merchantId, merchantId);
            ShippingSettings newSettings = SampleShippingSettings();
            UpdateShippingSettings(merchantId, merchantId, newSettings);
            GetShippingSettings(merchantId, merchantId);
            oldSettings.ETag = null; // Clear out the ETag first.
            UpdateShippingSettings(merchantId, merchantId, oldSettings);
            GetShippingSettings(merchantId, merchantId);
        }

        /// <summary>
        /// Retrieves the shipping settings for a particular account.
        /// </summary>
        /// <returns>The shipping settings for the specified account.</returns>
        private ShippingSettings GetShippingSettings(ulong merchantId, ulong accountId)
        {
            Console.WriteLine("=================================================================");
            Console.WriteLine("Getting Account Shipping Settings for {0}", accountId);
            Console.WriteLine("=================================================================");

            ShippingSettings settings = service.Shippingsettings.Get(merchantId, accountId).Execute();
            PrintShippingSettings(settings);
            Console.WriteLine();

            return settings;
        }

        private void UpdateShippingSettings(ulong merchantId, ulong accountId, ShippingSettings settings)
        {
            Console.WriteLine("=================================================================");
            Console.WriteLine("Updating Account Shipping Settings for {0}", accountId);
            Console.WriteLine("=================================================================");

            service.Shippingsettings.Update(settings, merchantId, accountId).Execute();
            Console.WriteLine();
        }

        private ShippingSettings SampleShippingSettings()
        {
            RateGroup rateUSPS = new RateGroup();
            rateUSPS.ApplicableShippingLabels = new List<string>();
            rateUSPS.SingleValue = new Value() {
              FlatRate = new Price() {
                Value = "5.00",
                Currency = "USD"
              }
            };

            Service serviceUSPS = new Service() {
              Name = "USPS",
              Currency = "USD",
              DeliveryCountry = "US",
              DeliveryTime = new DeliveryTime() {
                MinTransitTimeInDays = 3,
                MaxTransitTimeInDays = 7,
              },
              RateGroups = new List<RateGroup> { rateUSPS },
              Active = true
            };

            ShippingSettings settings = new ShippingSettings() {
              PostalCodeGroups = new List<PostalCodeGroup>(),
              Services = new List<Service> { serviceUSPS }
            };
            return settings;
        }

        private void PrintShippingSettings(ShippingSettings settings)
        {
            Console.WriteLine("Shipping settings for account {0}:", settings.AccountId);
            if (settings.PostalCodeGroups == null || settings.PostalCodeGroups.Count == 0)
            {
                Console.WriteLine("- No postal code groups.");
            }
            else
            {
                Console.WriteLine("- {0} postal code group(s):", settings.PostalCodeGroups.Count);
                foreach (var group in settings.PostalCodeGroups)
                {
                    Console.WriteLine("  Postal code group \"{0}\":", group.Name);
                    Console.WriteLine("  - Country: {0}", group.Country);
                    Console.WriteLine("  - Contains {0} postal code ranges.", group.PostalCodeRanges.Count);
                }
            }
            if (settings.Services == null || settings.Services.Count == 0)
            {
                Console.WriteLine("- No shipping services.");
            }
            else
            {
                Console.WriteLine("- {0} shipping service(s):", settings.Services.Count);
                foreach (var service in settings.Services)
                {
                    Console.WriteLine("  Service \"{0}\":", service.Name);
                    Console.WriteLine("  - Active: {0}", service.Active);
                    Console.WriteLine("  - Country: {0}", service.DeliveryCountry);
                    Console.WriteLine("  - Currency: {0}", service.Currency);
                    Console.WriteLine("  - Delivery time: {0} - {1} days",
                        service.DeliveryTime.MinTransitTimeInDays,
                        service.DeliveryTime.MaxTransitTimeInDays);
                    Console.WriteLine("  - {0} rate group(s) in this service.", service.RateGroups.Count);
                }
            }
        }
    }
}
